### PR TITLE
Implement tap operations

### LINKED ISSUE #48: Implement tap method
Tap is useful method to do sideeffect in the middle of chain for iterable collections (like logging). Eg:

```php
$ints
    ->map($power2)
    ->tap($log) // log in the middle
    ->map($multiply2)
```

It should have signature like

```php
/**
 * @param callable(T): void $sideEffect
 *
 * @return static<T>
 */
public function tap(callable $sideEffect): static
```

The semantic is, that tap **MUST NOT** mutate values inside of collection. It should either return the very same instance, or a new instance with same data (e.g. tap implemented on ArrayList can easily be implemented as each + return $this, and in that case, it can return the very same reference, but on LazyList, where the tap operation is deffered and does not materialize the collection, it should return a new reference of LazyList with contents unchanged.)

### PR DESCRIPTION
**Pull request by jost125**

### CODE REVIEW

File: src/Bonami/Collection/ArrayList.php [hunk 1]
```diff
@@ -810,6 +810,26 @@ public function each(callable $sideEffect): void
...
+     */
+    public function tapEach(callable $sideEffect)
+    {
+        foreach ($this->items as $key => $item) {
+            $sideEffect($item, $key);
+        }
+
+        return $this;
+    }
+
```
Comment (honzatrtik) [inline src/Bonami/Collection/ArrayList.php:832]: Im thinking about usefulness of `each` & `tapEach` together .. wdyt?

Comment (jost125) [reply src/Bonami/Collection/ArrayList.php:832]: They are both useful, as we concluded in discussion.

**Review by vasekboch — APPROVED** [review]

File: tests/Bonami/Collection/EitherTest.php [hunk 2]
```diff
@@ -238,6 +238,44 @@ public function testEach(): void
...
     }
 
+    public function testTap(): void
+    {
+        $right = Either::right(42);
+        $left = Either::left(666);
+        $accumulated = 0;
+
+        $accumulate = static function (int $i) use (&$accumulated): void {
+            $accumulated += $i;
```
Comment (honzatrtik) [inline tests/Bonami/Collection/EitherTest.php:248]: I would not like unexperienced devs to have impression, that this is the way, how to "fold" list, but they probably wont look at the test cases :-D

**Review by honzatrtik — APPROVED** [review]

File: tests/Bonami/Collection/EitherTest.php (hunk 2 above)
Comment (jost125) [reply tests/Bonami/Collection/EitherTest.php:248]: I had similar thoughts. But I hope, that we are targeting on more experienced audience. Also, there are some more obvious methods for "folding", like `sum`, `reduce`, which are obvious enough (mfold is not so obvious for unexperienced audience).

Anyway, we need to improve docs to help developers with choosing best practices.

**Pull request merged**